mm2src/coins/utxo/bch_and_slp_tx_history.rs

```diff
@@ -0,0 +1,410 @@
+                .await
+            {
+                return Self::change_state(Stopped::storage_error(e));
+            }
+
+            // wait for for one second to reduce the number of requests to electrum servers
+            Timer::sleep(1.).await;
+        }
+        info!("Tx history fetching finished for {}", ctx.coin.ticker());
+        *ctx.coin.as_ref().history_sync_state.lock().unwrap() = HistorySyncState::Finished;
+        Self::change_state(WaitForHistoryUpdateTrigger::new())
+    }
+}
+
+#[derive(Debug)]
+enum StopReason<E> {
+    HistoryTooLarge,
+    StorageError(E),
+    UnknownError(String),
+}
+
+struct Stopped<T, E> {
+    phantom: std::marker::PhantomData<T>,
+    stop_reason: StopReason<E>,
+}
+
+impl<T, E> Stopped<T, E> {
+    fn history_too_large() -> Self {
+        Stopped {
+            phantom: Default::default(),
+            stop_reason: StopReason::HistoryTooLarge,
+        }
+    }
+
+    fn storage_error(e: E) -> Self {
+        Stopped {
+            phantom: Default::default(),
+            stop_reason: StopReason::StorageError(e),
+        }
+    }
+
+    fn unknown(e: String) -> Self {
+        Stopped {
+            phantom: Default::default(),
+            stop_reason: StopReason::UnknownError(e),
+        }
+    }
+}
+
+impl<T, E> TransitionFrom<FetchingTxHashes<T>> for Stopped<T, E> {}
+impl<T, E> TransitionFrom<UpdatingUnconfirmedTxes<T>> for Stopped<T, E> {}
+impl<T, E> TransitionFrom<WaitForHistoryUpdateTrigger<T>> for Stopped<T, E> {}
+impl<T, E> TransitionFrom<FetchingTransactionsData<T>> for Stopped<T, E> {}
+
+#[async_trait]
+impl<T: TxHistoryStorage, E: std::fmt::Debug + Send + 'static> LastState for Stopped<T, E> {
+    type Ctx = BchAndSlpHistoryCtx<T>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> Self::Result {
+        info!(
+            "Stopping tx history fetching for {}. Reason: {:?}",
+            ctx.coin.ticker(),
+            self.stop_reason
+        );
+        let new_state_json = match self.stop_reason {
+            StopReason::HistoryTooLarge => json!({
+                "code": utxo_common::HISTORY_TOO_LARGE_ERR_CODE,
+                "message": "Got `history too large` error from Electrum server. History is not available",
+            }),
+            reason => json!({
+                "message": format!("{:?}", reason),
+            }),
+        };
+        *ctx.coin.as_ref().history_sync_state.lock().unwrap() = HistorySyncState::Error(new_state_json);
+    }
+}
+
+pub async fn bch_and_slp_history_loop(
+    coin: BchCoin,
+    storage: impl TxHistoryStorage,
+    metrics: MetricsArc,
+    current_balance: BigDecimal,
+) {
+    let ctx = BchAndSlpHistoryCtx {
+        coin,
+        storage,
+        metrics,
+        current_balance,
+    };
+    let state_machine: StateMachine<_, ()> = StateMachine::from_ctx(ctx);
+    state_machine.run(Init::new()).await;
```

Comment:
Do you think is it possible to reuse the states from this mod for other coins?

Reply:
Yes, I have left a doc-comment at the top of the module in regards to it. I plan to refactor all UTXO coins to use this state machine later - it would be too much for this PR.